File: server/engine/rtm.py
"""Direct port of src/lib/rtm.ts."""

from __future__ import annotations

from dataclasses import dataclass

from engine.models import MappingRow, TestCase


@dataclass
class RtmEntry:
    requirement_id: str
    mapping_row_id: str
    source_table: str
    source_field: str
    target_table: str
    target_field: str
    transformation: str
    is_primary_key: bool
    is_nullable: bool
    covered_test_case_ids: list[str]
    test_case_count: int
    covered: bool

    def to_json(self) -> dict:
        return {
            "requirementId": self.requirement_id,
            "mappingRowId": self.mapping_row_id,
            "sourceTable": self.source_table,
            "sourceField": self.source_field,
            "targetTable": self.target_table,
            "targetField": self.target_field,
            "transformation": self.transformation,
            "isPrimaryKey": self.is_primary_key,
            "isNullable": self.is_nullable,
            "coveredTestCaseIds": self.covered_test_case_ids,
            "testCaseCount": self.test_case_count,
            "covered": self.covered,
        }


def build_rtm(mapping_rows: list[MappingRow], test_cases: list[TestCase]) -> list[RtmEntry]:
    test_cases_by_mapping_row: dict[str, list[TestCase]] = {}
    for tc in test_cases:
        for row_id in tc.source_mapping_row_ids:
            test_cases_by_mapping_row.setdefault(row_id, []).append(tc)

    entries: list[RtmEntry] = []
    for index, row in enumerate(mapping_rows):
        covering = test_cases_by_mapping_row.get(row.id, [])
        entries.append(
            RtmEntry(
                requirement_id=f"REQ-{index + 1:03d}",
                mapping_row_id=row.id,
                source_table=row.source_table,
                source_field=row.source_field,
                target_table=row.target_table,
                target_field=row.target_field,
                transformation=row.transformation,
                is_primary_key=row.is_primary_key,
                is_nullable=row.is_nullable,
                covered_test_case_ids=[tc.id for tc in covering],
                test_case_count=len(covering),
                covered=len(covering) > 0,
            )
        )
    return entries

File: src/lib/rtm.ts
import type { MappingRow } from '@/types/mapping';
import type { TestCase, TestCategory } from '@/types/testCase';
import { TEST_CATEGORIES } from '@/types/testCase';

export interface RtmEntry {
  requirementId: string;
  mappingRowId: string;
  sourceTable: string;
  sourceField: string;
  targetTable: string;
  targetField: string;
  transformation: string;
  isPrimaryKey: boolean;
  isNullable: boolean;
  coveredTestCaseIds: string[];
  testCaseCount: number;
  covered: boolean;
}

/**
 * Builds a Requirements Traceability Matrix: one row per mapping requirement
 * (a source-to-target field mapping), linked to every generated test case that
 * covers it via TestCase.sourceMappingRowIds. A requirement with zero linked
 * test cases is a coverage gap — surfaced as covered: false.
 */
export function buildRtm(mappingRows: MappingRow[], testCases: TestCase[]): RtmEntry[] {
  const testCasesByMappingRow = new Map<string, TestCase[]>();
  for (const tc of testCases) {
    for (const rowId of tc.sourceMappingRowIds) {
      const list = testCasesByMappingRow.get(rowId) ?? [];
      list.push(tc);
      testCasesByMappingRow.set(rowId, list);
    }
  }

  return mappingRows.map((row, index) => {
    const covering = testCasesByMappingRow.get(row.id) ?? [];
    return {
      requirementId: `REQ-${String(index + 1).padStart(3, '0')}`,
      mappingRowId: row.id,
      sourceTable: row.sourceTable,
      sourceField: row.sourceField,
      targetTable: row.targetTable,
      targetField: row.targetField,
      transformation: row.transformation,
      isPrimaryKey: row.isPrimaryKey,
      isNullable: row.isNullable,
      coveredTestCaseIds: covering.map((tc) => tc.id),
      testCaseCount: covering.length,
      covered: covering.length > 0,
    };
  });
}

export interface CategoryRtmGroup {
  category: TestCategory;
  testCaseCount: number;
  coveredMappingRowCount: number;
  /** Every mapping row, with coverage fields (coveredTestCaseIds/testCaseCount/covered) scoped to
   *  just this category's test cases -- not the global "covered by anything" flag. */
  entries: RtmEntry[];
}

/**
 * Groups RTM coverage by test category instead of one row per mapping field -- a flat per-row
 * table gets unwieldy fast on a real mapping doc with dozens of fields across nine categories.
 * `requirementId` in buildRtm is derived only from mappingRows' own index, never from testCases,
 * so calling buildRtm once per category (scoped to that category's test cases) yields identical,
 * stable REQ-IDs across every group -- safe to call repeatedly.
 */
export function buildRtmByCategory(mappingRows: MappingRow[], testCases: TestCase[]): CategoryRtmGroup[] {
  const groups: CategoryRtmGroup[] = [];
  for (const category of TEST_CATEGORIES) {
    const categoryTestCases = testCases.filter((tc) => tc.category === category);
    if (categoryTestCases.length === 0) continue; // nothing to show for a category the user didn't generate
    const entries = buildRtm(mappingRows, categoryTestCases);
    groups.push({
      category,
      testCaseCount: categoryTestCases.length,
      coveredMappingRowCount: entries.filter((e) => e.covered).length,
      entries,
    });
  }
  return groups;
}
